File: frontend/src/api/collection.ts
import { apiFetch } from "./client";

export type CollectionEntry = {
  card_id: string;
  card_name: string;
  set_code: string;
  rarity: string;
  image_url: string;
  quantity: number;
  type_line?: string;
  oracle_text?: string;
  cmc?: number;
  colors: string[];
  set_icon?: string;
  scryfall_uri?: string;
};

export type CollectionQuery = {
  search?: string;
  color?: string;
  type_filter?: string;
  sort_by?: string;
};

export function getCollection(playerId: number, query: CollectionQuery = {}) {
  const params = new URLSearchParams();

  if (query.search) params.set("search", query.search);
  if (query.color) params.set("color", query.color);
  if (query.type_filter) params.set("type_filter", query.type_filter);
  if (query.sort_by) params.set("sort_by", query.sort_by);

  const suffix = params.toString() ? `?${params.toString()}` : "";
  return apiFetch<CollectionEntry[]>(`/collection/${playerId}${suffix}`);
}

export function addToCollection(playerId: number, cardId: string, quantity = 1) {
  return apiFetch(`/collection/${playerId}/add`, {
    method: "POST",
    body: JSON.stringify({
      card_id: cardId,
      quantity,
    }),
  });
}

export function removeFromCollection(playerId: number, cardId: string, quantity = 1) {
  return apiFetch(`/collection/${playerId}/remove`, {
    method: "POST",
    body: JSON.stringify({
      card_id: cardId,
      quantity,
    }),
  });
}

export type ImportedCardSearchResult = {
  id: string;
  name: string;
  set_code: string;
  rarity: string;
  type_line?: string;
  cmc?: number;
  colors: string[];
  image_url: string;
  scryfall_uri?: string;
  set_icon?: string;
};

export function searchImportedCards(query: string, limit = 20) {
  const params = new URLSearchParams();
  params.set("q", query);
  params.set("limit", String(limit));

  return apiFetch<ImportedCardSearchResult[]>(
    `/collection/search/cards?${params.toString()}`
  );
}

export function exportCollection(playerId: number) {
  return apiFetch(`/collection/${playerId}/export`);
}

export function importCollection(playerId: number, data: unknown) {
  return apiFetch(`/collection/${playerId}/import`, {
    method: "POST",
    body: JSON.stringify({ data }),
  });
}

File: backend/app/api/routes/collection.py
from fastapi import APIRouter, Query, HTTPException

from app.services.collection_service import get_collection

from pydantic import BaseModel
from app.services.collection_service import (
    get_collection,
    add_card_to_collection,
    remove_card_from_collection,
    export_collection,
    import_collection,
)

from app.services.card_repository import search_imported_cards

router = APIRouter(prefix="/collection", tags=["collection"])

class CollectionAdjustRequest(BaseModel):
    card_id: str
    quantity: int = 1

class CollectionImportRequest(BaseModel):
    data: dict

@router.get("/{player_id}")
def get_player_collection(
    player_id: int,
    search: str = Query(default=""),
    color: str = Query(default=""),
    type_filter: str = Query(default=""),
    sort_by: str = Query(default="name_asc"),
) -> list[dict]:
    return get_collection(
        player_id=player_id,
        search=search,
        color=color,
        type_filter=type_filter,
        sort_by=sort_by,
    )

@router.post("/{player_id}/add")
def add_to_collection(player_id: int, payload: CollectionAdjustRequest) -> dict:
    return add_card_to_collection(player_id, payload.card_id, payload.quantity)


@router.post("/{player_id}/remove")
def remove_from_collection(player_id: int, payload: CollectionAdjustRequest) -> dict:
    return remove_card_from_collection(player_id, payload.card_id, payload.quantity)

@router.get("/search/cards")
def search_cards(
    q: str = Query(default=""),
    limit: int = Query(default=20, ge=1, le=100),
) -> list[dict]:
    return search_imported_cards(q, limit)

@router.get("/{player_id}/export")
def export_player_collection(player_id: int) -> dict:
    try:
        return export_collection(player_id)
    except ValueError as exc:
        raise HTTPException(status_code=404, detail=str(exc)) from exc


@router.post("/{player_id}/import")
def import_player_collection(player_id: int, payload: CollectionImportRequest) -> dict:
    try:
        return import_collection(player_id, payload.data)
    except ValueError as exc:
        raise HTTPException(status_code=400, detail=str(exc)) from exc